README.md
# CWA Python Samples

[![Symantec](https://img.shields.io/badge/tag-symantec-yellow.svg)](https://www.symantec.com/)
[![CWA](https://img.shields.io/badge/tag-cwa-yellow.svg)](https://www.symantec.com/products/cloud-workload-protection)
[![Python](https://img.shields.io/badge/language-python-blue.svg)](https://www.python.org/)
|
[![GitHub contributors](https://img.shields.io/github/contributors/Symantec/cwa-python-samples.svg)](https://GitHub.com/Symantec/cwa-python-samples/graphs/contributors/)
|
[![GitHub issues](https://img.shields.io/github/issues/Symantec/cwa-python-samples.svg)](https://GitHub.com/Symantec/cwa-python-samples/issues/)
[![GitHub issues-closed](https://img.shields.io/github/issues-closed/Symantec/cwa-python-samples.svg)](https://GitHub.com/Symantec/cwa-python-samples/issues?q=is%3Aissue+is%3Aclosed)
[![GitHub pull-requests](https://img.shields.io/github/issues-pr/Symantec/cwa-python-samples.svg)](https://GitHub.com/Symantec/cwa-python-samples/pull/)

Python code samples for demonstrating usage of Symantec Cloud Workload Assurance (SCWA) RESTful API functions

-----------------------------------------------------------------------------------------------------------------------

- [CWA Python Samples](#cwa-python-samples)
  - [Postman](#postman)
  - [Setup](#setup)
  - [CHANGELOG](CHANGELOG.md)
  - [License](LICENSE)

**Scripts**

  - [Generate Auth Token](#Generate-Auth-Token)
  - [Add Account In Scan Profile](#Add-Account-In-Scan-Profile)
  - [Associate Account To Org Unit](#Associate-Account-To-Org-Unit)
  - [Create Child Org Unit Under Root OU](#Create-Child-Org-Unit-Under-Root-OU)
  - [Delete Org Unit](#Delete-Org-Unit)
  - [Get Accounts Associated With Org Unit](#Get-Accounts-Associated-With-Org-Unit)
  - [Get Events](#Get-Events)
  - [Get List Of Accounts](#Get-List-Of-Accounts)
  - [Get List Of Checks For Resoucres](#Get-List-Of-Checks-For-Resoucres)
  - [Get List Of Policies](#Get-List-Of-Policies)
  - [Get List Of Resources](#Get-List-Of-Resources)
  - [Get List Of Scan Profiles](#Get-List-Of-Scan-Profiles)
  - [Get List Of Services](#Get-List-Of-Services)
  - [Get Org Unit Hierarchy](#Get-Org-Unit-Hierarchy)
  - [Get Scan Status Using Scan Profile ID](#Get-Scan-Status-Using-Scan-Profile-ID)
  - [Start Scan Using Account And Policy IDs](#Start-Scan-Using-Account-And-Policy-IDs)
  - [Start Scan Using Scan ProfileID](#Start-Scan-Using-Scan-ProfileID)

-----------------------------------------------------------------------------------------------------------------------

## Postman

A Postman collection allows you to test the APIs, import this file.

Also create a new environment with the following keys and add your respective values:

`client_id` and `client_secret`.

- [CWA_REST_API_COLLECTION](CWA_REST_API_COLLECTION.postman_collection.json)

-----------------------------------------------------------------------------------------------------------------------

## Setup

Refer to Symantec CWA API documentation at: https://apidocs.symantec.com/home/SCWA

Before you get started you need a Symantec Cloud Workload Assurance Account. If you do not have one sign up for a trial account using this link, select the 'Cloud Workload Assurance' check box: https://securitycloud.symantec.com/cc/#/onboard

You can also buy Cloud Workload Assurance from Amazon AWS Market Place that also includes free usage. Click this link: https://aws.amazon.com/marketplace/pp/B07JM2CFK4

After you have activated your account, completed AWS or Azure Connection with periodic sync interval; you are ready to start using these samples.
First step is to Create API access keys. After login to CWA console, go to 'Settings' page and click on 'API Keys' tab

Copy following API secret keys for your CWA tenant ID information and secure them

`Customer ID: SEJ*#########################7788`

`Domain ID: Dq*####################6Yh`

`Client ID: O***#####################y988`

`Client Secret Key: t##################################`

-----------------------------------------------------------------------------------------------------------------------

**Code Files**

-----------------------------------------------------------------------------------------------------------------------
## Generate Auth Token
[cwa_generateAuthToken.py](cwa_generateAuthToken.py)

Script to generate Authentication token which will be be used for subsequent API calls.
Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_token_based_authentication_service

Usage: 
> `python cwa_generateAuthToken.py`

Sample:
>  `python cwa_generateAuthToken.py`

-----------------------------------------------------------------------------------------------------------------------
## Add Account In Scan Profile
[cwa_addAccountInScanProfile.py](cwa_addAccountInScanProfile.py)

This API service is used to add accounts in an existing scan profile. Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_addaaccountforprofile

Usage: 
> `python cwa_addAccountInScanProfile.py`

Sample:
>  `python cwa_addAccountInScanProfile.py`

-----------------------------------------------------------------------------------------------------------------------

## Associate Account To Org Unit
[cwa_associateAccountToOrgUnit.py](cwa_associateAccountToOrgUnit.py)

This API service is used to associate accounts to an organization unit. Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_addaaccountforprofile

Usage: 
> `python cwa_associateAccountToOrgUnit.py`

Sample:
>  `python cwa_associateAccountToOrgUnit.py`

-----------------------------------------------------------------------------------------------------------------------

## Create Child Org Unit Under Root OU
[cwa_createChildOrgUnitUnderRootOU.py](cwa_createChildOrgUnitUnderRootOU.py)

This API service is used to create an organization unit. Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_createchildorgunit

Usage: 
> `python cwa_createChildOrgUnitUnderRootOU.py`

Sample:
>  `python cwa_createChildOrgUnitUnderRootOU.py`

-----------------------------------------------------------------------------------------------------------------------

## Delete Org Unit
[cwa_deleteOrgUnit.py](cwa_deleteOrgUnit.py)

*Desc*

Usage: 
> `python cwa_deleteOrgUnit.py`

Sample:
>  `python cwa_deleteOrgUnit.py`

-----------------------------------------------------------------------------------------------------------------------

## Get Accounts Associated With Org Unit
[cwa_getAccountsAssociatedWithOrgUnit.py](cwa_getAccountsAssociatedWithOrgUnit.py)

This API service retrieves a list of accounts associated with an organization unit. Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_getassociatedaccounts

Usage: 
> `python cwa_getAccountsAssociatedWithOrgUnit.py`

Sample:
>  `python cwa_getAccountsAssociatedWithOrgUnit.py`

-----------------------------------------------------------------------------------------------------------------------

## Get Events
[cwa_getEvents.py](wa_getEvents.py)

Script to get events that are generated due to misconfigurations of AWS/Azure resources as per policy or checks.
Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_get_event_details

Usage: 
> `python cwa_getEvents.py`

Sample:
>  `python cwa_getEvents.py`

-----------------------------------------------------------------------------------------------------------------------
## Get List Of Accounts
[cwa_getListofAccounts.py](cwa_getListofAccounts.py)

Script to get the list of all available AWS and Azure accounts in the cloud infrastructure.
Refer to CWA REST API at https://apidocs.symantec.com/home/sCWA#_getaccounts

Usage: 
> `python cwa_getListofAccounts.py`

Sample:
>  `python cwa_getListofAccounts.py`

-----------------------------------------------------------------------------------------------------------------------
## Get List Of Checks For Resoucres
[cwa_getListofChecksForResoucres.py](cwa_getListofChecksForResoucres.py)

Script to get the list of checks along with the check details for the AWS and Azure resources' id you provide.
Refer to CWA REST API at https://apidocs.symantec.com/home/sCWA#_getchecksforservice

Usage: 
> `python cwa_getListofChecksForResoucres.py`

Sample:
>  `python cwa_getListofChecksForResoucres.py`

-----------------------------------------------------------------------------------------------------------------------
## Get List Of Policies
[cwa_getListofPolicies.py](cwa_getListofPolicies.py)

Script to get the list of all the existing policies. This includes policy information such as policy id and policy name.
Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_getpolicies

Usage: 
> `python cwa_getListofPolicies.py`

Sample:
>  `python cwa_getListofPolicies.py`

-----------------------------------------------------------------------------------------------------------------------
## Get List Of Resources
[cwa_getListofResources.py](cwa_getListofResources.py)

Script to get the list of resources for the service id you provide. Following are the services ID that are currently supported by CWA for AWS and Azure:

```
ServiceID:ServiceType:Cloud Provider

1:VPC:AWS
2:IAM Users:AWS
3:IAM Password Policy:AWS
4:VPC Security Groups:AWS
5:Simple Notification Service:AWS
6:CloudWatch:AWS
7:CloudTrail:AWS
8:IAM Encryption Keys:AWS
9:IAM:AWS
10:AWS Config:AWS
11:IAM Policies:AWS
12:S3:AWS
13:EC2 Elastic Block Store:AWS
14:EC2 Load Balancers:AWS
15:IAM and ACM Certificates:AWS
16:RDS:AWS
17:CloudFront:AWS
18:Amazon Redshift:AWS
19:RDS Event subscription:AWS
20:EC2:AWS
21:AMIs:AWS
22:AWS Account Settings:AWS
23:Simple Queue Service:AWS
24:VPC Network ACLs:AWS
25:VPC Subnets:AWS
26:CloudFormation:AWS
27:Virtual Networks:AZURE
28:Virtual Machines:AZURE
29:Subnets:AZURE
30:IAM Roles:AWS
```

Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_getserviceresources

Usage: 
> `python cwa_getListofResources.py`

Sample:
>  `python cwa_getListofResources.py`

-----------------------------------------------------------------------------------------------------------------------
## Get List Of Scan Profiles
[cwa_getListofScanProfiles.py](cwa_getListofScanProfiles.py)

Script to get the list of existing scan profiles.
Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_getscanprofiles 

Usage: 
> `python cwa_getListofScanProfiles.py`

Sample:
>  `python cwa_getListofScanProfiles.py`

-----------------------------------------------------------------------------------------------------------------------
## Get List Of Services
[cwa_getListofServices.py](cwa_getListofServices.py)

Script to get the list of available services along with service details such as service name, service id, region, and cloud provider.
Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_getservices 

Usage: 
> `python cwa_getListofServices.py`

Sample:
>  `python cwa_getListofServices.py`

-----------------------------------------------------------------------------------------------------------------------

## Get Org Unit Hierarchy
[cwa_getOrgUnitHierarchy.py](cwa_getOrgUnitHierarchy.py)

This API service retrieves the list of organization units with hierarchical structure. Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_getorgunithierarchy

Usage: 
> `python cwa_getOrgUnitHierarchy.py`

Sample:
>  `python cwa_getOrgUnitHierarchy.py`

-----------------------------------------------------------------------------------------------------------------------

## Get Scan Status Using Scan Profile ID
[cwa_getScanStatusUsingScanProfileID.py](cwa_getScanStatusUsingScanProfileID.py)

This API service retrieves the status of the job associated to the scan profile ID that you provide. Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_getjobstatus

Provide a `SCAN_PROFILE_ID`.

Usage: 
> `python cwa_getScanStatusUsingScanProfileID.py`

Sample:
>  `python cwa_getScanStatusUsingScanProfileID.py`

-----------------------------------------------------------------------------------------------------------------------
## Start Scan Using Account And Policy IDs
[cwa_startScanUsingAccountAndPolicyIDs.py](cwa_startScanUsingAccountAndPolicyIDs.py)

Script that lets you start a scan immediately for the given policy id and Account ID. The script returns the scan profile id.
Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_startscan 

Usage: 
> `python cwa_startScanUsingAccountAndPolicyIDs.py`

Sample:
>  `python cwa_startScanUsingAccountAndPolicyIDs.py`

-----------------------------------------------------------------------------------------------------------------------
## Start Scan Using Scan ProfileID
[cwa_startScanUsingScanProfileID.py](cwa_startScanUsingScanProfileID.py)

Script that lets you to start scan using a scan profile id.
Refer to CWA REST API at https://apidocs.symantec.com/home/SCWA#_startscanforprofile 

Usage: 
> `python cwa_startScanUsingScanProfileID.py`

Sample:
>  `python cwa_startScanUsingScanProfileID.py`

-----------------------------------------------------------------------------------------------------------------------

## Change Log

See [CHANGELOG](CHANGELOG.md).

-----------------------------------------------------------------------------------------------------------------------

## License

See [License](LICENSE).


CHANGELOG.md
# Change Log

6th July 2019 - [AlexHedley](https://github.com/alexhedley)
- Updated `README.md`

Nov 2018 - [nikkhilvijay](https://github.com/nikkhilvijay)
- Added scripts

30th Oct 2018 - [amitdwivedi21](https://github.com/amitdwivedi21)
- Added scripts

cwa_getListofResources.py
#!/usr/bin/env python 
#################################################################################################################
# Copyright 2018 Symantec Corporation. All rights reserved.
# Script to get Events
# You need to provide Client ID and Client Secret Key in config.ini file under Credentials sections, to generate
#  auth token.
# These keys can be found on CWA customer portal, once you are logged in navigate to Settings->API Keys
# The config file must be in the same directory as of the script.
#
# !!!Usage!!!: Update value of  variable SERVICE_ID to any value from 1 to 25 as per following mapping
# 1	VPC ; 2	IAM Users ; 3 IAM Password Policy ; 4 VPC Security Groups ; 5	Simple Notification Service
# 6	CloudWatch ; 7	CloudTrail ;8	IAM Encryption Keys ;9	IAM ;10	AWS Config; 11	IAM Policies ;12 S3
# 13 EC2 Elastic Block Store; 14	EC2 Load Balancers ;15	IAM and ACM Certificates ;16 RDS ;17 CloudFront
# 18 Amazon Redshift; 19	RDS Event subscription; 20	EC2 ;21	AMIs; 22 AWS Account Settings ;
#
# 23 Simple Queue Service ; 24 VPC Network ACLs; 25 VPC Subnets
# then execute script python
# python cwa_getListofResources.py
# On success, this script will produce resources of provided service type
#################################################################################################################
import json
import requests
import configparser
import os
import logging
from pathlib import Path

# create logger
logger = logging.getLogger("cwa_getListofResources")
logger.setLevel(logging.INFO)

# create console handler (ch) and set level to debug
ch = logging.StreamHandler()

# create formatter
formatter = logging.Formatter("%(asctime)s: %(levelname)s: %(message)s",
                              "%Y-%m-%d %H:%M:%S")
# add formatter to console handler
ch.setFormatter(formatter)

# add console handler to logger
logger.addHandler(ch)


# Setting variables from config.ini file
AUTHURL = 'AuthUrl'
CLIENT_ID = 'ClientId'
CLIENT_SECRET = 'ClientSecretKey'
CONFIG_CREDS_SECTION = 'Credentials'
GET_SERVICES_URL = 'GetServicesUrl'
CONFIG_URL_SECTION = 'RequestURL'
SERVICE_ID = '2' # Replace SERVICE_ID with desired


# Getting current working directory
Current_Working_Dir = os.getcwd()

logger.info("cwa_getListofResources: Current working Directory is " + Current_Working_Dir)
logger.info("cwa_getListofResources: Checking if config file present in current directory")

config_file = Path(Current_Working_Dir+'/config.ini')

# Checking if config file present
if not config_file.is_file():
    logger.error("cwa_getListofResources: File config.ini not found in current working directory, place config.ini "
                 "fle in directory \n " + Current_Working_Dir)
    exit()
else:
    logger.info("cwa_getListofResources: Configfile found in directory "+Current_Working_Dir)


# Creating http request and headers

auth_headers = {}
access_token = None
authurl = None
services_url = None
resources_rui = '/' + SERVICE_ID + '/resources'
x_epmp_customer_id = None
x_epmp_domain_id = None
get_resource_url = None


# Defining method to hit http request and to generate auth token


def get_authentication_token():
    token_cust_domain_id = False
    try:
        auth_request_json = json.dumps(set_request_headers.auth_request)
        logger.info("get_authentication_token(): Hitting http request to generate auth token ")
        auth_response = requests.post(authurl, data=auth_request_json, headers=auth_headers)
        if auth_response.status_code == 200:
            logger.info("get_authentication_token(): auth token generated successfully, "
                        "http status code is " + str(auth_response.status_code))
            global access_token
            access_token = auth_response.json()['access_token']
            global x_epmp_customer_id
            x_epmp_customer_id = auth_response.json()['x-epmp-customer-id']
            global x_epmp_domain_id
            x_epmp_domain_id = auth_response.json()['x-epmp-domain-id']
            print("access_token :: " + access_token)
            print("customer_id :: " + x_epmp_customer_id)
            print("domain_id :: " + x_epmp_domain_id)
            token_cust_domain_id = True
        else:
            logger.error("get_authentication_token(): Response from http auth not received status code is " + str(auth_response.status_code))

    except Exception as ex:
        logger.error("get_authentication_token(): Exception occurred while hitting http request to generate token" + str(ex))
    return token_cust_domain_id


def set_request_headers():
    set_request_headers.auth_request = {}
    headers_got_set = False
    try:
        logger.info("set_request_headers(): Parsing and reading values from config files")
        config = configparser.ConfigParser()
        config.read(config_file)
        client_id = config.get(CONFIG_CREDS_SECTION, CLIENT_ID)
        client_secret = config.get(CONFIG_CREDS_SECTION, CLIENT_SECRET)
        global authurl
        authurl = config.get(CONFIG_URL_SECTION, AUTHURL)
        global services_url
        services_url = config.get(CONFIG_URL_SECTION, GET_SERVICES_URL)
        if client_id == "" or client_secret == "" or authurl == "" or services_url == "":
            logger.error("set_request_headers(): One or more values empty in config_file")
            return headers_got_set
        else:
            global get_resource_url
            get_resource_url = services_url + resources_rui
            set_request_headers.auth_request['client_id'] = client_id
            set_request_headers.auth_request['client_secret'] = client_secret
            auth_headers['Content-type'] = 'application/json'
            headers_got_set = True

    except Exception as ex:
        logger.error("set_request_headers(): Exception occurred while reading values from config file " + str(ex))
    return headers_got_set


def get_list_of_resources():
    global auth_headers
    auth_headers['Authorization'] = access_token
    auth_headers['x-epmp-customer-id'] = x_epmp_customer_id
    auth_headers['x-epmp-domain-id'] = x_epmp_domain_id

    try:
        logger.info("get_list_of_resources(): Hitting http request to get list of resournces for "
                    "service id " + SERVICE_ID)
        resources_list_response = requests.get(get_resource_url, headers=auth_headers)
        if resources_list_response.status_code == 200:
            logger.info("get_list_of_resources(): List of resources fetched successfully "
                        + str(resources_list_response.status_code))
            resources_list_json_array = json.loads(resources_list_response.text)
            for resource in resources_list_json_array:
                resource_id = resource['resourceid']
                resource_name = resource['resourcename']
                print("resource id :: " + resource_id + "  resource name :: " + resource_name)
        else:
            logger.error("get_list_of_resources(): Failed to get lsit of resources, status code is "
                         + str(resources_list_response.status_code))
    except Exception as ex:
        logger.error("get_list_of_resources(): Exception occurred while retriving resources  " + str(ex))


if set_request_headers():
    if get_authentication_token():
        get_list_of_resources()



